module2/model/src/main/java/com/epam/esm/dao/CertificateDAOImpl.java

```diff
@@ -0,0 +1,204 @@
+import org.springframework.stereotype.Component;
+import java.sql.PreparedStatement;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Component
+public class CertificateDAOImpl implements CertificateDAO {
+    private final JdbcTemplate jdbcTemplate;
+    private final TagDAO tagDAO;
+    private final CertificateTagDAO certificateTagDAO;
+
+    private final String GET_ALL_SQL = "SELECT * FROM certificate";
+    private final String SAVE_SQL= "INSERT INTO certificate (name,description,price,duration,create_date,last_update_date) VALUES (?,?,?,?,?,?)";
+    private final String DELETE_SQL = "DELETE FROM certificate WHERE id=?";
+    private final String FIND_BY_ID_SQL = "SELECT * FROM certificate WHERE id=?";
+    private final String LIKE_SQL = "SELECT * FROM certificate WHERE name LIKE ? OR description LIKE ?";
+    private final String UPDATE_SQL = "UPDATE certificate SET name=?,description=?,price=?,duration=?,last_update_date=? WHERE id=?";
+    private final String ID_COLUMN= "id";
+    private final String NAME_COLUMN= "name";
+    private final String DESCRIPTION_COLUMN= "description";
+    private final String PRICE_COLUMN= "price";
+    private final String DURATION_COLUMN= "duration";
+
+    @Autowired
+    public CertificateDAOImpl(JdbcTemplate jdbcTemplate,
+                              TagDAO tagDAO,
+                              CertificateTagDAO certificateTagDAO) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.tagDAO = tagDAO;
+        this.certificateTagDAO = certificateTagDAO;
+    }
+
+    @Override
+    public List<Certificate> getAll(){
+        List<CertificateTag> list = certificateTagDAO.getAll();
+        List<Certificate> certificates = jdbcTemplate.query(GET_ALL_SQL,
+                new BeanPropertyRowMapper<>(Certificate.class));
+        return getCertificatesWithTags(list, certificates);
+    }
+
+    @Override
+    public void save(Certificate certificate) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(SAVE_SQL, new String[]{ID_COLUMN});
+            ps.setString(1, certificate.getName());
+            ps.setString(2, certificate.getDescription());
+            ps.setDouble(3, certificate.getPrice());
+            ps.setInt(4, certificate.getDuration());
+            ps.setString(5, ZonedDateTime.now(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE));
+            ps.setString(6, ZonedDateTime.now(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE));
+            return ps;
+        }, keyHolder);
+        Long certificateId = keyHolder.getKey().longValue();
+        this.saveTagsFromCertificate(certificate, certificateId);
+    }
+
+    @Override
+    public void update(Certificate certificate, int id) {
```

Comment:
I can see, that you are performing several operations in different tables here. But I can not see, where you use transactions

Reply:
This issue must be fixed for all cases over the project where transactions must applied.

Reply:
> This issue must be fixed for all cases over the project where transactions must applied.

add @Transactional